### PR TITLE
Distinguish host from world on ingress

### PR DESCRIPTION
**Pull request by joestringer**
Previously, all traffic that is masqueraded to the local host's IP would be considered as coming "from the host" for policy purposes, even if the traffic originally came from outside the node. This PR introduces a magic mark on packets sourced from the local stack to ensure that local traffic is considered as "host", while traffic from the outside world (or even other containers on the host) are considered as "world".

Related: #3940 

```release-note
Distinguish host from world on ingress.
Traffic from the outside world that would previously pass through an SNAT step would have
appeared (eg, in monitor output) as though it came from `reserved:host` (`host` entity).
It is now considered `reserved:world` (`world` entity), while only traffic that comes from 
a local process will be considered as `reserved:host`.
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/3988)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Hit a weird issue hit in CI, logs from `cilium-zvqqk-logs.log`:

```
2018-05-02T23:28:26.00831706Z level=info msg="Deleted CiliumNetworkPolicy" ciliumNetworkPolicyName=default-deny-all-ingress k8sApiVersion= k8sNamespace=default
2018-05-02T23:28:26.060388432Z level=info msg="Regenerating BPF program" containerID=59221c48c4 endpointID=13949 ipv4=10.10.1.32 ipv6="f00d::a0a:100:0:367d" k8sPodName="kube-system:microscope" policyRevision=34 startTime="2018-05-02 23:28:26.059875192 +0000 UTC m=+534.949845328"
2018-05-02T23:28:26.060585324Z level=info msg="Regenerating BPF program" containerID=746e10910f endpointID=173 ipv4=10.10.1.66 ipv6="f00d::a0a:100:0:ad" k8sPodName="kube-system:kube-dns-1992960195-bxz47" policyRevision=34 startTime="2018-05-02 23:28:26.060501731 +0000 UTC m=+534.950471849"
2018-05-02T23:28:26.062844995Z level=info msg="Regenerating BPF program" containerID=cilium-loc endpointID=29898 ipv4=10.10.1.172 ipv6="f00d::a0a:100:0:74ca" k8sPodName=":" policyRevision=34 startTime="2018-05-02 23:28:26.062675859 +0000 UTC m=+534.952645975"
2018-05-02T23:31:04.022773306Z level=warning msg="Cilium API handler panicked" client=@ method=GET panic_message="write unix /var/run/cilium/cilium.sock->@: write: broken pipe" url=/v1/healthz
2018-05-02T23:31:04.097043332Z level=warning msg="Cilium API handler panicked" client=@ method=GET panic_message="write unix /var/run/cilium/cilium.sock->@: write: broken pipe" url=/v1/healthz
2018-05-02T23:31:04.102507157Z level=warning msg="Cilium API handler panicked" client=@ method=GET panic_message="write unix /var/run/cilium/cilium.sock->@: write: broken pipe" url=/v1/healthz
2018-05-02T23:31:04.863310473Z level=error msg="Error while getting a new service ID. Ignoring service..." error="Error while locking path cilium-net/operational/ServicesV2/SHA256SUMServices/5bcc14a1c1e088876339488e63efded680dc86372aabf83826a03bb672bba414: etcdserver: requested lease not found" ipAddr=10.96.0.10 k8sNamespace=kube-system k8sSvcName=kube-dns port=53 protocol=TCP serviceID=dns-tcp
2018-05-02T23:32:22.267009324Z level=error msg="error while receiving request from xDS stream" error="rpc error: code = 1 desc = context canceled" xdsStreamID=3
2018-05-02T23:32:22.2694891Z level=error msg="Failed to retrieve the container list" error="Cannot connect to the Docker daemon. Is the docker daemon running on this host?" subsys=containerd-watcher
2018-05-02T23:32:22.269509092Z level=info msg="Envoy: Closing access log connection"
2018-05-02T23:32:22.36637077Z level=error msg="xDS stream context canceled" error="context canceled" xdsStreamID=1
2018-05-02T23:32:22.370627981Z level=error msg="error while receiving request from xDS stream" error="rpc error: code = 1 desc = context canceled" xdsStreamID=1
2018-05-02T23:32:22.370710518Z level=error msg="xDS stream context canceled" error="context canceled" xdsStreamID=2
2018-05-02T23:32:22.390243493Z level=error msg="error while receiving request from xDS stream" error="rpc error: code = 1 desc = context canceled" xdsStreamID=2
2018-05-02T23:32:22.392945095Z level=error msg="xDS stream context canceled" error="context canceled" xdsStreamID=3
2018-05-02T23:32:22.608390695Z level=warning msg="Envoy: Proxy crashed" error="signal: killed"
2018-05-02T23:32:22.980051012Z level=warning msg="Cilium API handler panicked" client=@ method=GET panic_message="write unix /var/run/cilium/cilium.sock->@: write: broken pipe" url=/v1/healthz
2018-05-02T23:32:23.081647837Z level=info msg="Envoy: Proxy started with pid 2812"
2018-05-02T23:32:23.081687498Z level=info msg="Envoy: Sleeping for 100ms before restarting proxy"
2018-05-02T23:32:23.085693536Z level=warning msg="Cilium API handler panicked" client=@ method=GET panic_message="write unix /var/run/cilium/cilium.sock->@: write: broken pipe" url=/v1/healthz
2018-05-02T23:32:23.862103365Z level=error msg="Cannot update CEP" containerID=cilium-loc controller="sync-to-k8s-ciliumendpoint (29898)" endpointID=29898 error="Operation cannot be fulfilled on ciliumendpoints.cilium.io \"cilium-health-k8s2\": StorageError: invalid object, Code: 4, Key: /registry/cilium.io/ciliumendpoints/kube-system/cilium-health-k8s2, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 68798083-4e60-11e8-a1ec-0800277280d1, UID in object meta: " ipv4=10.10.1.172 ipv6="f00d::a0a:100:0:74ca" k8sPodName=":" policyRevision=0
```

**Comment by joestringer** [comment]
Seems like the above issue is unrelated, splitting out into #3994 . Will retry tests.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

File: daemon/daemon.go [hunk 1]
```diff
@@ -581,6 +588,20 @@ func (d *Daemon) installIptablesRules() error {
 		return err
 	}
 
+	// Mark all packets sourced from the host with a special marker so that
```
Comment (tgraf) [inline daemon/daemon.go]: Please be more specific here and state that we mark packets from *processes* running on the host.

**Review by tgraf — DISMISSED** [review]

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Hit this in CI, seems fairly reliable on this PR but doesn't happen locally:

### Stacktrace

```
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:312
Unexpected failure pinging host (192.168.254.254) from httpd2: PING 192.168.254.254 (192.168.254.254): 56 data bytes
--- 192.168.254.254 ping statistics ---
5 packets transmitted, 0 packets received, 100% packet loss

Expected
    <bool>: false
to be true
/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/runtime/Policies.go:628
```

### Standard Output
```
�[1mSTEP�[0m: Setting PolicyEnforcement=default
time="2018-05-15T00:16:56Z" level=info msg="setting PolicyEnforcement=default"
time="2018-05-15T00:16:56Z" level=debug msg="running command: sudo cilium config -o json | jq -r '.status.realized[\"policy-enforcement\"]'"
time="2018-05-15T00:16:56Z" level=debug msg="running command: sudo cilium config PolicyEnforcement=default"
time="2018-05-15T00:17:01Z" level=debug msg="running command: cilium endpoint list -o jsonpath='{range [*]}{@.status.external-identifiers.container-name}{\"=\"}{@.status.state}{\"\\n\"}{end}'"
time="2018-05-15T00:17:02Z" level=info msg="'7' containers are in a 'ready' state of a total of '7' containers." functionName=WaitEndpointsReady status="map[ready:7]" test=RunPolicies
time="2018-05-15T00:17:02Z" level=info msg="VM IP address configuration" IPv4_host=192.168.254.254 IPv4_other_host=192.168.254.111 IPv4_other_net=99.11.0.0/16 IPv6_host="fdff::ff" test=RunPolicies
time="2018-05-15T00:17:02Z" level=debug msg="running command: sudo cilium bpf endpoint delete 192.168.254.254"
time="2018-05-15T00:17:02Z" level=debug msg="running command: sudo cilium bpf endpoint delete fdff::ff"
time="2018-05-15T00:17:02Z" level=debug msg="running command: sudo docker inspect httpd1"
�[1mSTEP�[0m: Getting IPv4 and IPv6 prefixes of node-local endpoints
time="2018-05-15T00:17:02Z" level=debug msg="running command: expr 10.15.13.37 : '\\([0-9]*\\.[0-9]*\\.\\)'"
time="2018-05-15T00:17:02Z" level=debug msg="running command: expr 10.15.13.37 : '\\([0-9]*\\.[0-9]*\\.\\)'"
�[1mSTEP�[0m: IPV6 Prefix: f00d::a0f:0:0:8ad6/112
�[1mSTEP�[0m: IPV4 Address Endpoint: 10.15.13.37
�[1mSTEP�[0m: IPV4 Prefix: 10.15.0.0/16
�[1mSTEP�[0m: IPV4 Prefix Except: 10.15.0.0/18
�[1mSTEP�[0m: Setting PolicyEnforcement to always enforce (default-deny)
�[1mSTEP�[0m: Setting PolicyEnforcement=always
time="2018-05-15T00:17:02Z" level=info msg="setting PolicyEnforcement=always"
time="2018-05-15T00:17:02Z" level=debug msg="running command: sudo cilium config -o json | jq -r '.status.realized[\"policy-enforcement\"]'"
time="2018-05-15T00:17:02Z" level=debug msg="running command: sudo cilium config PolicyEnforcement=always"
time="2018-05-15T00:17:08Z" level=debug msg="running command: cilium endpoint list -o jsonpath='{range [*]}{@.status.external-identifiers.container-name}{\"=\"}{@.status.state}{\"\\n\"}{end}'"
time="2018-05-15T00:17:08Z" level=info msg="'7' containers are in a 'ready' state of a total of '7' containers." functionName=WaitEndpointsReady status="map[ready:7]" test=RunPolicies
�[1mSTEP�[0m: Adding Pseudo-Host IPs to localhost
time="2018-05-15T00:17:08Z" level=debug msg="running command: sudo ip addr add dev lo 192.168.254.254/32"
time="2018-05-15T00:17:08Z" level=debug msg="running command: sudo ip addr add dev lo fdff::ff/128"
�[1mSTEP�[0m: Pinging host IPv4 from httpd2 (should NOT work due to default-deny PolicyEnforcement mode)
time="2018-05-15T00:17:08Z" level=debug msg="running command: docker exec -i  httpd2 ping -W 2 -c 5 192.168.254.254"
�[1mSTEP�[0m: Importing L3 CIDR Policy for IPv4 Egress Allowing Egress to 192.168.254.111, 192.168.254.111 from id.httpd2
time="2018-05-15T00:17:15Z" level=debug msg="PolicyRenderAndImport: render policy to 'policy_afd3a30c.json'" test=RunPolicies
time="2018-05-15T00:17:15Z" level=debug msg="PolicyRenderAndImport: import policy from '/home/vagrant/go/src/github.com/cilium/cilium/test/policy_afd3a30c.json'" test=RunPolicies
�[1mSTEP�[0m: Setting up policy: /home/vagrant/go/src/github.com/cilium/cilium/test/policy_afd3a30c.json
time="2018-05-15T00:17:15Z" level=debug msg="running command: sudo cilium policy get -o json | jq '.revision'"
time="2018-05-15T00:17:15Z" level=info msg="before importing policy" file-path=/home/vagrant/go/src/github.com/cilium/cilium/test/policy_afd3a30c.json policyRevision=20 test=RunPolicies
time="2018-05-15T00:17:15Z" level=info msg="validating policy before importing" file-path=/home/vagrant/go/src/github.com/cilium/cilium/test/policy_afd3a30c.json test=RunPolicies
time="2018-05-15T00:17:15Z" level=debug msg="running command: sudo cilium policy validate /home/vagrant/go/src/github.com/cilium/cilium/test/policy_afd3a30c.json"
time="2018-05-15T00:17:16Z" level=debug msg="running command: sudo cilium policy import /home/vagrant/go/src/github.com/cilium/cilium/test/policy_afd3a30c.json"
time="2018-05-15T00:17:21Z" level=debug msg="running command: sudo cilium policy get -o json | jq '.revision'"
time="2018-05-15T00:17:21Z" level=debug msg="running command: sudo cilium policy wait 23"
time="2018-05-15T00:17:21Z" level=debug msg="running command: sudo cilium policy get -o json | jq '.revision'"
time="2018-05-15T00:17:21Z" level=info msg="policy import finished and revision increased" file-path=/home/vagrant/go/src/github.com/cilium/cilium/test/policy_afd3a30c.json policyRevision=23 test=RunPolicies
time="2018-05-15T00:17:21Z" level=debug msg="running command: docker exec -i  httpd2 ping -W 2 -c 5 192.168.254.254"
time="2018-05-15T00:17:28Z" level=debug msg="running command: sudo ip addr del dev lo 192.168.254.254/32"
```

Weirdly, I don't see the relevant CIDRs in the ipcache:

```
IP PREFIX/ADDRESS        IDENTITY                                               
f00d::a0f:0:0:ad/128     43713                                                  
f00d::a0f:0:0:74ca/128   13812                                                  
10.15.242.54/32          13812                                                  
0.0.0.0/0                2                                                      
10.15.42.252/32          31830                                                  
10.15.13.37/32           933                                                    
10.15.28.238/32          1                                                      
10.15.250.189/32         7505                                                   
10.15.193.100/32         20106                                                  
f00d::a0f:0:0:8ad6/128   1                                                      
f00d::/64                3                                                      
f00d::a0f:0:0:981b/128   7505                                                   
10.0.0.0/8               3                                                      
f00d::a0f:0:0:dc06/128   20984                                                  
10.15.148.173/32         43713                                                  
::/0                     2                                                      
fc00::10ca:1/128         1                                                      
f00d::a0f:0:0:367d/128   31830                                                  
10.15.118.183/32         27473                                                  
10.0.2.15/32             1                                                      
10.15.129.91/32          20984                                                  
f00d::a0f:0:0:bf00/128   933                                                    
f00d::a0f:0:0:72c5/128   20106      
```
```
$ grep 192 cilium-bugtool-20180515-001728.989+0000-UTC-076773416/cmd/cilium-bpf-ipcache-list.md
$
```

[test_results_Cilium-PR-Ginkgo-Tests-Validated_3174.zip](https://github.com/cilium/cilium/files/2003725/test_results_Cilium-PR-Ginkgo-Tests-Validated_3174.zip)

Seems like the ipIdentityWatcher hasn't been triggered for the past 8(!) minutes:

```
RuntimeValidatedPoliciesChecksCIDRL3Policy $ grep -B 1 ipcache cilium-bugtool-20180515-001728.989+0000-UTC-076773416/cmd/gops-stack-\$\(pidof-cilium-agent\).md 
goroutine 401 [chan receive, 8 minutes]:
github.com/cilium/cilium/pkg/ipcache.ipIdentityWatcher(0xc421f7ea80, 0x2, 0x2)
        /home/vagrant/go/src/github.com/cilium/cilium/pkg/ipcache/ipcache.go:508 +0x139
created by github.com/cilium/cilium/pkg/ipcache.InitIPIdentityWatcher.func1
        /home/vagrant/go/src/github.com/cilium/cilium/pkg/ipcache/ipcache.go:643 +0x50
```

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Attempting a CI run with just the IPTables rule change to attempt to bisect the above issue.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Successful CI run with the first patch. Rebasing and putting back the whole branch.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Latest failure:
```
Error creating policy "/home/vagrant/go/src/github.com/cilium/cilium/test/k8sT/manifests/cnp-specs.yaml"
Expected
    <*errors.errorString | 0xc420b61300>: {
        s: "cannot get the revision ",
    }
to be nil
```

Which is hitting in this code:
https://github.com/cilium/cilium/blob/bf4794429f2e819af5353640589cf0319dc10904/test/helpers/kubectl.go#L869-L873

Found in logs:

```
2018-05-17T23:44:03.293217476Z level=debug msg="Queueing container event" containerID=aaef8ea407 event=exec_die subsys=containerd-watcher
2018-05-17T23:44:03.296378938Z level=debug msg="Queueing container event" containerID=aaef8ea407 event="exec_create: cilium status" subsys=containerd-watcher
2018-05-17T23:44:03.296402734Z level=debug msg="Queueing container event" containerID=aaef8ea407 event="exec_create: cilium status" subsys=containerd-watcher
2018-05-17T23:44:03.296406775Z level=debug msg="Queueing container event" containerID=aaef8ea407 event="exec_start: cilium status" subsys=containerd-watcher
2018-05-17T23:44:03.296409937Z level=debug msg="Queueing container event" containerID=aaef8ea407 event="exec_start: cilium status" subsys=containerd-watcher
2018-05-17T23:44:03.298747218Z level=debug msg="Connecting to k8s to retrieve labels for pod in ns" k8sNamespace=default k8sPodName=details-v1-4288466782-2s9jj subsys=containerd-watcher                                                                      
2018-05-17T23:44:03.357227956Z level=warning msg="Cilium API handler panicked" client=@ method=GET panic_message="write unix /var/run/cilium/cilium.sock->@: write: broken pipe" url=/v1/healthz
2018-05-17T23:44:03.366764496Z level=warning msg="Cilium API handler panicked" client=@ method=GET panic_message="write unix /var/run/cilium/cilium.sock->@: write: broken pipe" url=/v1/healthz
```

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Test passed. Will run again.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Tests passed again. I'll drop the apipanic patch and run again. If it's still good, then I'm satisfied that the PR is not introducing any flakes.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Test passed: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/3415/

File: daemon/daemon.go [hunk 2]
```diff
@@ -537,12 +544,13 @@ func (d *Daemon) installIptablesRules() error {
 	// performed by kube-proxy for all packets destined for Cilium. Cilium
 	// installs a dedicated rule which does the source PAT to the right
 	// source IP.
+	clearMasqBit := fmt.Sprintf("%#04x/%#04x", 0, proxy.MagicMarkK8sMasq)
```
Comment (rlenglet) [inline daemon/daemon.go]: The option to `--set-mark` is 32-bit, and is dumped as such in iptables output. So I'd replace `%#04x` with `%#08x` to be consistent with that format.

Same for all the others below.

**Review by rlenglet — APPROVED** [review]
The proxy bits LGTM.
Just one nitpick about the formatting of marks / masks.

**Comment by joestringer** [comment]
test-me-please

File: daemon/daemon.go (hunk 2 above)
Comment (joestringer) [reply daemon/daemon.go]: Added an extra commit to clean this up.

**Review by borkmann — APPROVED** [review]
BPF bits LGTM

**Review by aanm — APPROVED** [review]

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
I didn't rebase bindata correctly. This time it should work.

**Comment by joestringer** [comment]
test-me-please

**Comment by michi-covalent** [comment]
shouldn't this be considered compatibility breaking change?

**Comment by ianvernon** [comment]
@michi-covalent in what sense? Can you please clarify? Are you saying that there might be a policy decision that will change from allow --> deny because of this?

**Comment by michi-covalent** [comment]
@ianvernon yes existing policies will start behaving differently right?

**Comment by ianvernon** [comment]
@michi-covalent I'm not entirely sure :( deferring to @joestringer on this one.

**Comment by joestringer** [comment]
True, the classification of traffic from the outside world when in this SNAT mode will switch from host to world. Perhaps we need to document this better.

**Pull request merged**